indexing-service/src/main/java/io/druid/indexing/common/task/IndexTask.java

```diff
@@ -431,41 +629,38 @@ public SegmentIdentifier allocate(DateTime timestamp, String sequenceName, Strin
             }
 
             final Interval interval = optInterval.get();
-            if (!shardSpecLookups.containsKey(interval)) {
-              final List<ShardSpec> intervalShardSpecs = shardSpecs.get(interval);
-              if (intervalShardSpecs == null || intervalShardSpecs.isEmpty()) {
-                throw new ISE("Failed to get shardSpec for interval[%s]", interval);
-              }
-              shardSpecLookups.put(interval, intervalShardSpecs.get(0).getLookup(intervalShardSpecs));
-            }
-
-            final ShardSpec shardSpec = shardSpecLookups.get(interval)
-                                                        .getShardSpec(inputRow.getTimestampFromEpoch(), inputRow);
-
-            final String sequenceName = String.format("index_%s_%s_%d", interval, version, shardSpec.getPartitionNum());
-
-            if (!sequenceNameToShardSpecMap.containsKey(sequenceName)) {
-              final ShardSpec shardSpecForPublishing = ingestionSchema.getTuningConfig().isForceExtendableShardSpecs()
-                                                       || ingestionSchema.getIOConfig().isAppendToExisting()
-                                                       ? new NumberedShardSpec(
-                  shardSpec.getPartitionNum(),
-                  shardSpecs.get(interval).size()
-              )
-                                                       : shardSpec;
-
-              sequenceNameToShardSpecMap.put(sequenceName, shardSpecForPublishing);
-            }
-
+            final ShardSpec shardSpec = shardSpecs.getShardSpec(interval, inputRow);
+            final String sequenceName = Appenderators.getSequenceName(interval, version, shardSpec);
             final AppenderatorDriverAddResult addResult = driver.add(inputRow, sequenceName, committerSupplier);
 
-            if (!addResult.isOk()) {
-              throw new ISE("Could not allocate segment for row with timestamp[%s]", inputRow.getTimestamp());
+            if (addResult.isOk()) {
+              if (addResult.getNumRowsInSegment() >= maxRowsInSegment ||
```

Comment:
Does this need a check to avoid the early publish if forceGuaranteedRollup=true?